File: miniprogram/src/domain/common/RuleReference.ts
export interface RuleReference {
  readonly ruleVersion: string
  readonly templateVersion: string
  readonly contentVersion: string
}

export function createRuleReference(
  ruleVersion: string,
  templateVersion: string,
  contentVersion: string,
): RuleReference {
  for (const [name, value] of Object.entries({
    ruleVersion,
    templateVersion,
    contentVersion,
  })) {
    if (value.trim() === '') {
      throw new Error(`${name} must not be blank`)
    }
  }

  return { ruleVersion, templateVersion, contentVersion }
}

File: backend/src/main/java/com/aifitness/assistant/common/domain/RuleReference.java
package com.aifitness.assistant.common.domain;

public record RuleReference(String ruleVersion, String templateVersion, String contentVersion) {

    public RuleReference {
        ruleVersion = requireText(ruleVersion, "ruleVersion");
        templateVersion = requireText(templateVersion, "templateVersion");
        contentVersion = requireText(contentVersion, "contentVersion");
    }

    private static String requireText(String value, String name) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(name + " must not be blank");
        }
        return value;
    }
}
